pkg/controller/cronjob/utils.go

```diff
@@ -121,33 +121,87 @@ func getRecentUnmetScheduleTimes(sj batchv1beta1.CronJob, now time.Time) ([]time
 	}
+
 	return starts, nil
 }
 
+// getLatestMissedSchedule returns the latest start time in the time window (inclusive), and the number of schedules in the time window (capped at 2).
+// The number of missed start times conveys "none, one, multiple".
+func getLatestMissedSchedule(startWindow time.Time, endWindow time.Time, schedule cron.Schedule) (time.Time, int) {
+	nextSchedule := schedule.Next(startWindow)
+
+	// If no schedules in window, return.
+	if nextSchedule.After(endWindow) {
+		return time.Time{}, 0
+
+		// There is (at least one) schedule in the window. See if there are more later ones.
+	} else {
+		latestSchedule, found := getLatestMissedScheduleBinarySearch(nextSchedule, endWindow, schedule)
+		if found && latestSchedule != nextSchedule {
+			return latestSchedule, 2 // We missed at least 2 schedules. Return the latest.
+		} else {
+			return nextSchedule, 1
+		}
+	}
+}
+
+// LatestScheduleTime, foundSchedule
+// getLatestMissedSchedule returns the latest start time in the specified time window (inclusive), and a true/false indicating if a start time was found.
+// It uses a binary search with the cron.Next function, to avoid stepping through every cron schedule.
+// This is less efficient than if cron.Previous existed, but is arguably simpler to review and test than a custom Previous function.
+func getLatestMissedScheduleBinarySearch(startWindow time.Time, endWindow time.Time, schedule cron.Schedule) (time.Time, bool) {
+	// Crons can't schedule in lower granularity than a minute.
+	// If we are looking at a time window less than a minute, there is "room" for at most 1 schedule, and we can skip the binary search.
+	minimumTimeUnit := time.Minute
+	if endWindow.Sub(startWindow) < minimumTimeUnit {
+		nextStart := schedule.Next(startWindow)
+		if !nextStart.After(endWindow) {
+			return nextStart, true
+		} else {
+			return time.Time{}, false
+		}
+	}
+
+	// Break the window into 2 halves for a binary search.
+	midway := startWindow.Add(endWindow.Sub(startWindow) / 2)
+
+	// Check the second half of the window for a start time.
+	// If there is a start time within this window, we can safely ignore the first half of the window.
+	nextScheduleAfterMidway := schedule.Next(midway)
+	if !nextScheduleAfterMidway.After(endWindow) {
+		latestFound, found := getLatestMissedScheduleBinarySearch(nextScheduleAfterMidway, endWindow, schedule) // Check a sub-window between the found schedule and the end of the window.
+		if found {
+			return latestFound, true
+		} else {
+			return nextScheduleAfterMidway, true
+		}
+
+		// If there was no start time in the second half of the window, we need to check the first half of the window.
+	} else {
+
+		nextScheduleInWindow := schedule.Next(startWindow)
+		if !nextScheduleInWindow.After(midway) { // Latest start time is in first half. There's no start time in the second half.
+			latestFound, found := getLatestMissedScheduleBinarySearch(startWindow, midway, schedule) // Check the first half.
```

Comment:
> The underlying implementation of Next() is sensitive to the duration of time between the start and the result

If that's the case, do you also want:
```suggestion
			latestFound, found := getLatestMissedScheduleBinarySearch(nextScheduleInWindow, midway, schedule) // Check the first half.
```